File: day6_part1_2023.rs
use std::fs;

fn calculate_ways_to_win(time: i32, record: i32) -> i32 {
    let mut ways_to_win = 0;
    for hold_time in 1..time {
        let travel_time = time - hold_time;
        let distance = hold_time * travel_time;
        if distance > record {
            ways_to_win += 1;
        }
    }
    ways_to_win
}

fn main() {
    let contents = fs::read_to_string("input.txt")
        .expect("Error reading file");

    let mut times: Vec<i32> = Vec::new();
    let mut distances: Vec<i32> = Vec::new();

    let lines: Vec<&str> = contents.trim().split("\n").collect();
    for (i, line) in lines.iter().enumerate() {
        let values: Vec<i32> = line.split(" ")
            .filter_map(|x| x.parse().ok())
            .collect();
        if i == 0 {
            times = values;
        } else {
            distances = values;
        }
    }

    let mut total_ways = 1;
    for i in 0..times.len() {
        let ways = calculate_ways_to_win(times[i], distances[i]);
        total_ways *= ways;
    }

    println!("{}", total_ways);
}

File: objectivec/day6_part1_2023.m
#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        // Read input from file
        NSString *filePath = @"input.txt";
        NSString *fileContent = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:nil];

        // Split the input into lines
        NSArray<NSString *> *lines = [fileContent componentsSeparatedByString:@"\n"];

        // Extract times and distances
        NSArray<NSString *> *timeParts = [[lines[0] componentsSeparatedByString:@":"][1] componentsSeparatedByString:@" "];
        NSArray<NSString *> *distanceParts = [[lines[1] componentsSeparatedByString:@":"][1] componentsSeparatedByString:@" "];

        NSMutableArray<NSNumber *> *times = [NSMutableArray array];
        NSMutableArray<NSNumber *> *distances = [NSMutableArray array];

        for (NSString *part in timeParts) {
            if (![part isEqualToString:@""]) {
                [times addObject:@([part intValue])];
            }
        }

        for (NSString *part in distanceParts) {
            if (![part isEqualToString:@""]) {
                [distances addObject:@([part intValue])];
            }
        }

        // Calculate the number of ways to beat the record for each race
        int product = 1;
        for (NSUInteger i = 0; i < times.count; i++) {
            int time = [times[i] intValue];
            int recordDistance = [distances[i] intValue];

            int waysToWin = 0;
            for (int holdTime = 0; holdTime <= time; holdTime++) {
                int travelTime = time - holdTime;
                int distance = holdTime * travelTime;
                if (distance > recordDistance) {
                    waysToWin++;
                }
            }
            product *= waysToWin;
        }

        // Print the result
        printf("%d\n", product);
    }
    return 0;
}